Analysis/Tasks/PWGCF/dptdptcorrelations.cxx

```diff
@@ -0,0 +1,407 @@
+#include "Framework/runDataProcessing.h"
+#include "Framework/AnalysisTask.h"
+#include "Framework/AnalysisDataModel.h"
+#include "Framework/ASoAHelpers.h"
+#include "Analysis/EventSelection.h"
+#include "Analysis/Centrality.h"
+#include "Analysis/TrackSelection.h"
+#include "Analysis/TrackSelectionTables.h"
+#include <TROOT.h>
+#include <TList.h>
+#include <TDirectory.h>
+#include <TH1.h>
+#include <TH2.h>
+#include <TH3.h>
+#include <TProfile3D.h>
+
+#include <cmath>
+
+using namespace o2;
+using namespace o2::framework;
+using namespace o2::soa;
+using namespace o2::framework::expressions;
+
+namespace o2
+{
+namespace aod
+{
+using CollisionEvSelCent = soa::Join<aod::Collisions, aod::EvSels, aod::Cents>::iterator;
+using TrackData = soa::Join<aod::Tracks, aod::TracksCov, aod::TracksExtra, aod::TracksExtended, aod::TrackSelection>::iterator;
+
+/* we have to change from int to bool when bool columns work properly */
+namespace dptdptcorrelations
+{
+DECLARE_SOA_COLUMN(EventAccepted, eventaccepted, uint8_t);
+DECLARE_SOA_COLUMN(TrackacceptedAsOne, trackacceptedasone, uint8_t);
+DECLARE_SOA_COLUMN(TrackacceptedAsTwo, trackacceptedastwo, uint8_t);
+} // namespace dptdptcorrelations
+DECLARE_SOA_TABLE(AcceptedEvents, "AOD", "ACCEPTEDEVENTS", dptdptcorrelations::EventAccepted);
+DECLARE_SOA_TABLE(ScannedTracks, "AOD", "SCANNEDTRACKS", dptdptcorrelations::TrackacceptedAsOne, dptdptcorrelations::TrackacceptedAsTwo);
+} // namespace aod
+} // namespace o2
+
+namespace dptdptcorrelations
+{
+/* all this bins have to be make configurable */
+int ptbins = 18;
+float ptlow = 0.2, ptup = 2.0;
+int etabins = 16;
+float etalow = -0.8, etaup = 0.8;
+int zvtxbins = 40;
+float zvtxlow = -10.0, zvtxup = 10.0;
+int phibins = 72;
+std::string fTaskConfigurationString = "PendingToConfigure";
+
+Configurable<int> cfgTrackType{"trktype", 1, "Type of selected tracks: 0 = no selection, 1 = global tracks loose DCA, 2 = global SDD tracks"};
+Configurable<int> cfgTrackOneCharge("trk1charge", 1, "Trakc one charge: 1 = positive, -1 = negative");
+Configurable<int> cfgTrackTwoCharge("trk2charge", -1, "Trakc two charge: 1 = positive, -1 = negative");
+Configurable<bool> cfgProcessPairs("processpairs", true, "Process pairs: false = no, just singles, true = yes, process pairs");
+
+/* while we don't have proper bool columnst */
+uint8_t DPTDPT_TRUE = 1;
+uint8_t DPTDPT_FALSE = 0;
+
+/// \enum SystemType
+/// \brief The type of the system under analysis
+enum SystemType {
+  kNoSystem = 0, ///< no system defined
+  kpp,           ///< **p-p** system
+  kpPb,          ///< **p-Pb** system
+  kPbp,          ///< **Pb-p** system
+  kPbPb,         ///< **Pb-Pb** system
+  kXeXe,         ///< **Xe-Xe** system
+  knSystems      ///< number of handled systems
+};
+
+/* probably this will not work in the context of distributed processes */
+/* we need to revisit it!!!!                                           */
+SystemType fSystem = kNoSystem;
+TH1F* fhCentMultB = nullptr;
```

Comment:
What is the reason to use these static pointers? They are not shared between the tasks, since those run in separate processes, each one will have their own.

Reply:
Perhaps I misunderstood something here. Do you mean that the OutputObject that are defined at the task level will be shared by all instances of the same task in a parallel running scenario with each task receiving a different input data file?

Reply:
OutputObj has a smart pointer managed by framework, I see that most of the histograms you have can be statically declared directly as OutputObj, which already behave as pointers and have correct ownership. 

Each "timeframe" or input file will result in a separate OutputObj for a given task, of course, they are not directly shared and their contents is automatically merged each time task finished processing a file.  I'm just wondering why do you need these pointers separately? Inside process you can always do `auto obj = *outputobj;` if you need the contained object itself, or you can call that contained object methods with `outputobj->method();`.

Reply:
Ok! That's what I thought.
Probably this global pointers will end up within the task scope. I have them here global because before I had two tasks one before event selection and another one after event selection which filled a different set of histograms.

But I definitely wanted them configured at task init time because the histograms will be configuration dependent. For instance, if the task is run for pp the histograms will be different as it is run for PbPb or if it is run for MC reco level or MC gen level. I want to have the histogram construction depending on the configuration, or even the running conditions, on the init section of the task and not 'hard' coded.

This also reminds me the limitation in the number of histograms which will be fixed with the HistogramRegistry.
If a TList can be used as an OutputObj we could use any number of histograms not declaring them as OutputObj just appending them to the TList as far a we are sure we are using the same list structure along all task instances. I think that was one of the things I was trying when I realized that a TList cannot be used as an OutputObj

Reply:
Tlist can be used with OutputObj! There is an example in histograms.cxx.